src/extension.ts

```diff
@@ -28,40 +28,54 @@ export async function activate(context: vscode.ExtensionContext) {
 
 	await activateSourceKitLSP(context);
 
-	// Check if we have a workspace folder open.
-	// This only supports single-root workspaces.
-	let workspaceRoot: string | undefined;
-	if (vscode.workspace.workspaceFolders) {
-		workspaceRoot = vscode.workspace.workspaceFolders[0].uri.fsPath;
-	}
-
-	// Features past this point require a workspace folder.
-	if (!workspaceRoot) {
-		return;
-	}
+	let workspaceContext = new WorkspaceContext(context);
+	let onWorkspaceChange = vscode.workspace.onDidChangeWorkspaceFolders((event) => {
+		if (workspaceContext === undefined) { console.log("Trying to run onDidChangeWorkspaceFolders on deleted context"); return; }
+		workspaceContext.onDidChangeWorkspaceFolders(event);
+	});
 
-	let ctx = await SwiftContext.create(workspaceRoot, context);
+	// Register commands.
+	const taskProvider = vscode.tasks.registerTaskProvider('swift', new SwiftTaskProvider(workspaceContext));
+	commands.register(workspaceContext);
 
-	// Register tasks and commands.
-	const taskProvider = vscode.tasks.registerTaskProvider('swift', new SwiftTaskProvider(ctx));
-	commands.register(ctx);
+	// observer for logging workspace folder addition/removal
+	let logObserver = workspaceContext.observerFolders((folder, operation) => {
+		console.log(`${operation}: ${folder.rootFolder.uri.fsPath}`);
+	});
 
-	// Create the Package Dependencies view.
-	const dependenciesProvider = new PackageDependenciesProvider(ctx);
-	const dependenciesView = vscode.window.createTreeView('packageDependencies', {
-		treeDataProvider: dependenciesProvider,
-		showCollapseAll: true
+	// observer that will add dependency view based on whether a root workspace folder has been added
+	let addDependencyViewObserver = workspaceContext.observerFolders((folder, operation) => {
+		if (folder.isRootFolder && operation === 'add') {
+			const dependenciesProvider = new PackageDependenciesProvider(folder);
+			const dependenciesView = vscode.window.createTreeView('packageDependencies', {
+				treeDataProvider: dependenciesProvider,
+				showCollapseAll: true
+			});
+			context.subscriptions.push(dependenciesView);
+		}
 	});
 
-	// Watch for changes to Package.swift and Package.resolved.
-	const packageWatcher = new PackageWatcher(workspaceRoot, ctx);
-	packageWatcher.install();
+	// observer that will resolve package for root folder
+	let resolvePackageObserver = workspaceContext.observerFolders(async (folder, operation) => {
+		if (folder.isRootFolder && operation === 'add') {
+			// resolve root package
+			if (folder.isRootFolder) {
+				// Create launch.json files based on package description. 
+				await debug.makeDebugConfigurations(folder);
+				await commands.resolveDependencies();
+			}
+		}
+	});
 
-	// Initialize the context keys and trigger a resolve task if needed.
-	packageWatcher.handlePackageSwiftChange();
+	// add workspace folders, already loaded
+	if (vscode.workspace.workspaceFolders !== undefined) {
```

Comment:
```suggestion
	if (vscode.workspace.workspaceFolders) {
```

That's quite common in JS/TS. It checks for  both `null` and `undefined`.

Reply:
ok